packages/base/src/blockchain.ts

```diff
@@ -278,3 +296,121 @@ export const CellbaseWitness = table(
+ *  outputsData: HexString[];
+ *  version: HexNumber;
+ *  witnesses: HexString[];
+ *}
+ * to :
+ * interface TransactionCodecType {
+ *   raw: {
+ *     version: Uint32LE;
+ *     cellDeps: DeCellDepVec;
+ *     headerDeps: Byte32Vec;
+ *     inputs: CellInputVec;
+ *     outputs: CellOutputVec;
+ *     outputsData: BytesVec;
+ *   };
+ *   witnesses: BytesVec;
+ * }
+ * ```
+ * @param data Transantion defined in @ckb-lumos/base/lib/api.d.ts
+ * @returns TransactionCodecType
+ */
+export function transformTransactionCodecType(
+  data: api.Transaction
+): TransactionCodecType {
+  return {
+    raw: {
+      version: data.version,
+      cellDeps: data.cellDeps,
+      headerDeps: data.headerDeps,
+      inputs: data.inputs,
+      outputs: data.outputs,
+      outputsData: data.outputsData,
+    },
+    witnesses: data.witnesses,
+  };
+}
+
+export function deTransformTransactionCodecType(
+  data: TransactionUnpackResultType
+): api.Transaction {
+  return {
+    cellDeps: data.raw.cellDeps.map((cellDep) => {
+      return {
+        outPoint: {
+          txHash: cellDep.outPoint.txHash,
+          index: BI.from(cellDep.outPoint.index).toHexString(),
+        },
+        depType: cellDep.depType,
+      };
+    }),
+    headerDeps: data.raw.headerDeps,
+    inputs: data.raw.inputs.map((input) => {
+      return {
+        previousOutput: {
+          txHash: input.previousOutput.txHash,
+          index: BI.from(input.previousOutput.index).toHexString(),
+        },
+        since: input.since.toHexString(),
+      };
+    }),
+    outputs: data.raw.outputs.map((output) => {
+      return {
+        capacity: output.capacity.toHexString(),
+        lock: output.lock,
+        type: output.type,
+      };
+    }),
+    outputsData: data.raw.outputsData,
+    version: BI.from(data.raw.version).toHexString(),
+    witnesses: data.witnesses,
+  };
+}
+
+export function transformHeaderCodecType(data: api.Header): HeaderCodecType {
+  return {
+    raw: {
+      timestamp: data.timestamp,
+      number: data.number,
+      epoch: data.epoch,
+      compactTarget: Number(data.compactTarget),
+      dao: data.dao,
+      parentHash: data.parentHash,
+      proposalsHash: data.proposalsHash,
+      transactionsRoot: data.transactionsRoot,
+      extraHash: data.extraHash,
+      version: data.version,
+    },
+    nonce: data.nonce,
+  };
+}
+
+export function deTransformHeaderCodecType(
+  data: HeaderUnpackResultType
+): api.Header {
+  return {
+    timestamp: data.raw.timestamp.toHexString(),
+    number: data.raw.number.toHexString(),
+    epoch: data.raw.epoch.toHexString(),
+    compactTarget: BI.from(data.raw.compactTarget).toHexString(),
+    dao: data.raw.dao,
+    parentHash: data.raw.parentHash,
+    proposalsHash: data.raw.proposalsHash,
+    transactionsRoot: data.raw.transactionsRoot,
+    extraHash: data.raw.extraHash,
+    version: BI.from(data.raw.version).toHexString(),
+    nonce: data.nonce.toHexString(),
+    hash: "",
```

Comment:
why empty hash here?

Reply:
The `hash` won't be packed whatever it is

Reply:
so we should remove the hash?